### PR TITLE
Agg pushdown v2

### PR DESCRIPTION
**Pull request by gruuya**
- Enable pushdown of qual (`WHERE`) clauses in combination with aggregations.
- Additionally, enable pushdown of `COUNT(*)`

CU-1z461e4

### CODE REVIEW

File: src/multicorn.c [hunk 1]
```diff
@@ -581,7 +572,24 @@ multicornGetForeignPlan(PlannerInfo *root,
...
+         * serialized.
+         */
         multicorn_extract_upper_rel_info(root, fdw_scan_tlist, planstate);
+
+        /*
+         * Since scan_clauses are empty in case of upper relations for some
+         * reason. We pass the clauses from the base relation obtained in MulticornGetForeignRelSize.
+         */
+        ofpinfo = (MulticornPlanState *) planstate->outerrel->fdw_private;
+        planstate->baserestrictinfo = extract_actual_clauses(ofpinfo->baserestrictinfo, false);
```
Comment (mildbyte) [inline src/multicorn.c:592]: Just to check my understanding:

  - This code used to split `baserestrictinfo` into `local_conds` and `remote_conds` with `multicorn_classify_conditions` in `multicornGetForeignRelSize`
  - Instead, we add the full `baserestrictinfo` to the plan state and use it to `extractRestrictions` into the `qual_list` (that makes it to Python)

observations/questions:
  - as far as I can see, `remote_conds` isn't actually used anywhere on master, so removing it has no effect
  - `local_conds` is used here (https://github.com/splitgraph/Multicorn/blob/cbe45ed682dde713aa46a08205ee794c1d6f948a/src/deparse.c#L497) and here (https://github.com/splitgraph/Multicorn/blob/cbe45ed682dde713aa46a08205ee794c1d6f948a/src/multicorn.c#L1686-L1692) but it looks like we never actually populate it after this change? Does that affect anything?

File: src/deparse.c [hunk 2]
```diff
@@ -239,11 +212,20 @@ multicorn_foreign_expr_walker(Node *node,
 				if (schema != PG_CATALOG_NAMESPACE)
 					return false;
 
-				/* Make sure the specific function at hand is shippable
+                /* Make sure the specific function at hand is shippable
                  * NB: here we deviate from standard FDW code, since the allowed
                  * function list is fetched from the Python FDW instance
                  */
-				if (!list_member(fpinfo->agg_functions, makeString(opername)))
+                if (agg->aggstar)
```
[the remark this answers is not in the document]
Comment (gruuya) [reply src/deparse.c:219]: Good point bringing `COUNT(1)` up. I see this doesn't get pushed down, so I'll need to investigate a bit (and neither does the equivalent `SELECT COUNT(es.account.*) FROM es.account`) though they would be used rarely I imagine.

However note that `COUNT(some_col)` [is not equivalent to](https://www.postgresql.org/docs/12/functions-aggregate.html) `COUNT(*)`, as the former discards `NULL` values while the latter takes them into account.

File: src/multicorn.c [hunk 3]
```diff
@@ -641,30 +651,50 @@ multicornBeginForeignScan(ForeignScanState *node, int eflags)
...
 		execstate->tupdesc = multicorn_get_tupdesc_for_join_scan_tuples(node);
 #else
 		execstate->tupdesc = node->ss.ss_ScanTupleSlot->tts_tupleDescriptor;
 #endif
         initConversioninfo(execstate->cinfos, TupleDescGetAttInMetadata(execstate->tupdesc), execstate->upper_rel_targets);
+
+        /*
+         * In case of a join or aggregate use the lowest-numbered member RTE out
+         * of all all the base relations participating in the underlying scan.
+         * NB: This may not work well in case of joins, keep an eye out for it.
```
[the remark this answers is not in the document]
Comment (gruuya) [reply src/multicorn.c:683]: This is adapted from `postgres_fdw` ([here](https://github.com/postgres/postgres/blob/REL_12_STABLE/contrib/postgres_fdw/postgres_fdw.c#L1425-L1434)), where it is being used for a different purpose.

We do have a [test with self-joins](https://github.com/splitgraph/splitgraph/blob/multicorn-agg-pushdown-v2-cu-1z461e4/test/splitgraph/commands/test_es_aggregation_pushdown.py#L552) in the engine repo, these queries are not pushed down.

File: src/python.c [hunk 4]
```diff
@@ -944,7 +944,7 @@ execute(ForeignScanState *node, ExplainState *es)
 		}
 		if (newqual != NULL)
 		{
-			PyObject   *python_qual = qualdefToPython((MulticornConstQual *) newqual, state->cinfos);
+			PyObject   *python_qual = qualdefToPython((MulticornConstQual *) newqual, state->qual_cinfos);
```
[the remark this answers is not in the document]
Comment (gruuya) [reply src/python.c:947]: It doesn't influence the existing behavior, since for non-agg queries `state->cinfos` is equal to `state->qual_cinfos`, given that [original table's tuple metadata](https://github.com/splitgraph/Multicorn/blob/agg-pushdown-v2-cu-1z461e4/src/multicorn.c#L2035) for non-agg queries matches the [target list tuple metadata](https://github.com/splitgraph/Multicorn/blob/agg-pushdown-v2-cu-1z461e4/src/multicorn.c#L658), which is not the case for aggregations. In short, I've added the `state->qual_cinfos` in order to track tuple metadata for the original table (i.e. columns of table in `FROM`) once the target tuple metadata (i.e. items in `SELECT` clause) deviates from it, which is the case in aggregations.

Note that regular `cinfos` are still being used for parsing the result of `execute` [here](https://github.com/splitgraph/Multicorn/blob/agg-pushdown-v2-cu-1z461e4/src/multicorn.c#L1117).

File: src/multicorn.c (hunk 1 above)
Comment (gruuya) [reply src/multicorn.c:592]: Your understanding is correct. And we don't populate `local_conds` so this change basically has no effect.

Effectively, I took the liberty of treating all `baserestrictinfo` as `remote_conds` for now, since `local_conds` come into play with joins, which are not pushed down presently.

File: src/deparse.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (gruuya) [reply src/deparse.c:219]: Ok, the reason `COUNT(1)` doesn't get pushed down is because `1` is treated as an expression, in particular a node of type `Const` which currently is [not being parsed](https://github.com/splitgraph/Multicorn/blob/agg-pushdown-v2-cu-1z461e4/src/deparse.c#L138). So full record fetch is performed and the counting is done on PG side.

I could add a case for parsing this node type now which is not much work in itself, but potentially some work handling edge cases since this increases the pushdown surface so to speak. Alternatively, I can just add a test for `COUNT(1)` as a reminder to have that handled at some point later on.

**Comment by gruuya** [comment]
I see one more issue here, and that is that PG isn't rechecking quals for pushed aggregations:
```sql
sgr@localhost:splitgraph> SELECT count(*) FROM es.account WHERE firstname ~~ 'Su_an%'
+-------+
| count |
|-------|
| 4     |
+-------+
SELECT 1
Time: 0.019s
sgr@localhost:splitgraph> SELECT count(*) FROM es.account WHERE firstname !~~ 'Su_an%'
+-------+
| count |
|-------|
| 1000  |
+-------+
SELECT 1
Time: 0.013s
sgr@localhost:splitgraph> select count(1) FROM es.account WHERE firstname !~~ 'Su_an%'
+-------+
| count |
|-------|
| 996   |
+-------+
SELECT 1
Time: 0.032s
```

File: src/deparse.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mildbyte) [reply src/deparse.c:219]: Ah, yep, `COUNT(some_col)` makes sense. Adding a test for COUNT(1) works.

File: src/multicorn.c (hunk 1 above)
Comment (mildbyte) [reply src/multicorn.c:592]: Got it -- is it worth deleting that code that uses `local_conds` altogether (since those two statements I linked now look like no-ops to me) and removing it from the struct?

**Comment by gruuya** [comment]
Ok, I think it's clear to me now after skimming PG source - [quals are not meant](https://github.com/postgres/postgres/blob/REL_12_STABLE/contrib/postgres_fdw/postgres_fdw.c#L1289) to be rechecked in case of pushed down aggregations, because if FDW says the pushdown is safe, there's no way to disentangle the returned aggregate result and discard the contributing rows which do not match quals.

Instead, what needs to be done is to reject aggregation pushdown in the plan phase if the qual operator is not supported,[here](https://github.com/pgspider/sqlite_fdw/blob/master/deparse.c#L615) is an example for SQLite FDW.

File: src/multicorn.c (hunk 1 above)
Comment (gruuya) [reply src/multicorn.c]: Looks like with the new approach I'll need the second if statement, need to still rethink it a bit.

Comment (gruuya) [reply src/multicorn.c:592]: Ok, I've decided to go with a much simpler implementation in the end, and so was able to remove the `local_conds` completely.

File: src/python.c [hunk 5]
```diff
@@ -1769,6 +1770,31 @@ canPushdownUpperrel(MulticornPlanState * state)
 		}
         Py_XDECREF(p_object);
 
+        /* Construct supported qual operators list */
+        p_ops = PyMapping_GetItemString(p_upperrel_pushdown, "operators_supported");
```
Comment (mildbyte) [inline src/python.c:1774]: Minor: this code looks very similar to the one that takes `agg_functions.keys()` and copies it into the `state->agg_functions` list just above; is it possible to refactor both into something like this?

```
void appendPySequenceToList(PyObject* p_object, List* target)
```

**Review by mildbyte — APPROVED** [review]

File: src/python.c (hunk 5 above)
Comment (gruuya) [reply src/python.c:1774]: Good catch, I'll add that function.
